pkg/ratelimiter/rate_limiter_test.go

```diff
@@ -114,3 +97,67 @@ func TestSpendConcurrent(t *testing.T) {
 	entry := rl.fillAndReturnEntry(PUBLISH, walletAddress, false)
 	require.Equal(t, entry.tokens, uint16(0))
 }
+
+func TestBucketExpiration(t *testing.T) {
+	// Set things up so that entries are expired after two sweep intervals
+	expiresAfter := 100 * time.Millisecond
+	sweepInterval := 60 * time.Millisecond
+
+	logger, _ := zap.NewDevelopment()
+	rl := NewTokenBucketRateLimiter(logger)
+	rl.Limits[DEFAULT] = &Limit{2, 0} // 2 tokens, no refill
+
+	require.NoError(t, rl.Spend(DEFAULT, "ip1", 1, false)) // bucket1 add
+	require.NoError(t, rl.Spend(DEFAULT, "ip2", 1, false)) // bucket1 add
+
+	time.Sleep(sweepInterval)
+	require.Equal(t, 0, rl.sweepAndSwap(expiresAfter)) // sweep bucket2 and swap
+
+	require.NoError(t, rl.Spend(DEFAULT, "ip2", 1, false)) // bucket1 refresh
+	require.NoError(t, rl.Spend(DEFAULT, "ip3", 1, false)) // bucket2 add
+
+	time.Sleep(sweepInterval)
+	require.Equal(t, 1, rl.sweepAndSwap(expiresAfter)) // sweep bucket1 and swap, delete ip1
+
+	// ip2 has been refreshed every 60ms so it should still be out of tokens
+	require.Error(t, rl.Spend(DEFAULT, "ip2", 1, false)) // bucket1 refresh
+	// ip1 entry should have expired by now, so we should have 2 tokens again
+	require.NoError(t, rl.Spend(DEFAULT, "ip1", 1, false)) // bucket1 add
+	require.NoError(t, rl.Spend(DEFAULT, "ip1", 1, false)) // bucket1 refresh
+	require.Error(t, rl.Spend(DEFAULT, "ip1", 1, false))   // bucket1 refresh
+
+	time.Sleep(sweepInterval)
+	require.Equal(t, 1, rl.sweepAndSwap(expiresAfter)) // sweep bucket2 and swap, delete ip3
+
+	// ip2 should still be out of tokens
+	require.Error(t, rl.Spend(DEFAULT, "ip2", 1, false)) // bucket1 refresh
+	// ip3 should have expired now and we should have 2 tokens again
+	require.NoError(t, rl.Spend(DEFAULT, "ip3", 1, false)) // bucket2 add
+	require.NoError(t, rl.Spend(DEFAULT, "ip3", 1, false)) // bucket2 refresh
+	require.Error(t, rl.Spend(DEFAULT, "ip3", 1, false))   // bucket2 refresh
+}
+
+func TestBucketExpirationIntegrity(t *testing.T) {
```

Comment:
@neekolas is this good enough? I dropped the janitor goroutine and instead am invoking the sweepAndSwap directly for better determinism. This is arguably covered by ip2 in the prior test, but I'm adding this to focus on a single specific entry.